includes/AMP/Output_Buffer.php

```diff
@@ -0,0 +1,295 @@
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Process response to ensure AMP validity.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param string $response HTML document response. By default it expects a complete document.
+	 * @return string AMP document response.
+	 */
+	public function prepare_response( string $response ): string {
+		// Enforce UTF-8 encoding as it is a requirement for AMP.
+		if ( ! headers_sent() ) {
+			header( 'Content-Type: text/html; charset=utf-8' );
+		}
+
+		/**
+		 * Document.
+		 *
+		 * @var Document $dom Document.
+		 */
+		$dom = Document::fromHtml( $response );
+
+		if ( ! $dom instanceof Document ) {
+			return $this->render_error_page( new Exception() );
+		}
+
+		// Make sure scripts from the body get moved to the head.
+		// TODO: Move to its own sanitizer?
+
+		/**
+		 * Scripts in <body>.
+		 *
+		 * @var DOMElement[] $scripts List of scripts.
+		 */
+		$scripts = $dom->xpath->query( '//body//script[ @custom-element or @custom-template or @src = "https://cdn.ampproject.org/v0.js" ]' );
+
+		foreach ( $scripts as $script ) {
+			if ( $script->parentNode ) {
+				$dom->head->appendChild( $script->parentNode->removeChild( $script ) );
+			}
+		}
+
+		$this->sanitize_document( $dom );
+		$this->optimize_document( $dom );
+
+		return $dom->saveHTML();
+	}
+
+
+	/**
+	 * Render error page.
+	 *
+	 * @todo Improve error message.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param Throwable $throwable Exception or (as of PHP7) Error.
+	 * @return string Error page.
+	 */
+	private function render_error_page( Throwable $throwable ): string {
+		return esc_html__( 'There was an error generating the web story, probably because of a server misconfiguration. Try contacting your hosting provider or open a new support request.', 'web-stories' );
+	}
+
+	/**
+	 * Sanitizes a document to be valid AMP.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param Document $document Document instance.
+	 *
+	 * @return void
+	 */
+	protected function sanitize_document( Document $document ) {
+		$sanitization = new Sanitization();
+		$sanitization->sanitize_document( $document );
+	}
+
+	/**
+	 * Optimizes AMP document.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param Document $document Document instance.
+	 *
+	 * @return void
+	 */
+	protected function optimize_document( Document $document ) {
+		$optimization = new Optimization();
+		$optimization->optimize_document( $document );
+	}
+
+	/**
+	 * Filters the Web Stories AMP sanitizers.
+	 *
+	 * @since 1.10.0
+	 *
+	 * @param array $sanitizers Sanitizers.
+	 * @return array Sanitizers.
+	 */
+	public function add_web_stories_amp_content_sanitizers( $sanitizers ): array {
+		if ( ! is_singular( 'web-story' ) ) {
+			return $sanitizers;
+		}
+
+		$post = get_queried_object();
+
+		if ( ! ( $post instanceof WP_Post ) ) {
+			return $sanitizers;
+		}
+
+		$video_cache_enabled = $this->experiments->is_experiment_enabled( 'videoCache' ) && (bool) get_option( Settings::SETTING_NAME_VIDEO_CACHE );
+
+		$story = new Story();
```

Comment:
Should this be injected as well?

Reply:
I don't see how that would be possible?

Reply:
Looking at this again, we should rather move this whole `add_web_stories_amp_content_sanitizers` method out of this class as it doesn't really belong here.

There's really no reason why this sanitizer is added via a filter. It can be added directly in `\Google\Web_Stories\AMP\Sanitization::get_sanitizers` like our other custom sanitizers.